numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1120,6 +1120,30 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
         goto fail;
     }
 
+    if (!PyArray_CanCastArrayTo(indices, dtype, NPY_SAFE_CASTING)) {
+        if (PyArray_SIZE(indices) == 0) {
+            /*
+             * Unfortunately, if the input array is empty, then indices would
+             * likely have a default dtype of float64, which isn't compatibile.
+             * Force a recast of an empty array.
+             */
+            PyArrayObject *indices_tmp = NULL;
+            Py_INCREF(dtype);
+            indices_tmp = (PyArrayObject*)PyArray_FromArray(indices, dtype,
+                                                            NPY_ARRAY_FORCECAST);
```

Comment:
Note, I am not being flippant. I totally get your concerns. That is why I posted this PR initially as a strawman. If you want to suggest a different approach, I would be more than happy to implement it, but my bosses want this problem fixed one way or another.